### PR TITLE
Add first class Analytics Engine bindings

### PR DESCRIPTION
**Pull request by jimhawkridge**
I've tried to add support for an `analytics_engine_datasets` section in wrangler.toml. To avoid having to use unsafe bindings.

### CODE REVIEW

File: packages/wrangler/src/config/validation.ts [hunk 1]
```diff
@@ -1956,6 +1968,43 @@ const validateServiceBinding: ValidatorFn = (diagnostics, field, value) => {
...
+	}
+	let isValid = true;
+	// Service bindings must have a binding and optional dataset.
+	if (!isRequiredProperty(value, "binding", "string")) {
+		diagnostics.errors.push(
+			`"${field}" bindings should have a string "binding" field but got ${JSON.stringify(
+				value
+			)}.`
+		);
+		isValid = false;
```
Comment (penalosa) [inline packages/wrangler/src/config/validation.ts:1992]: Can this just be `return false`?

File: packages/wrangler/src/config/validation.ts [hunk 2]
```diff
@@ -1956,6 +1968,43 @@ const validateServiceBinding: ValidatorFn = (diagnostics, field, value) => {
...
+	if (
+		!isOptionalProperty(value, "dataset", "string") ||
+		(value as { dataset: string }).dataset?.length === 0
+	) {
+		diagnostics.errors.push(
+			`"${field}" bindings should, optionally, have a string "dataset" field but got ${JSON.stringify(
+				value
+			)}.`
+		);
+		isValid = false;
```
Comment (penalosa) [inline packages/wrangler/src/config/validation.ts:2003]: As above

File: packages/wrangler/src/config/index.ts [hunk 3]
```diff
@@ -211,6 +212,21 @@ export function printBindings(bindings: CfWorkerInit["bindings"]) {
...
+	if (
+		analytics_engine_datasets !== undefined &&
+		analytics_engine_datasets.length > 0
+	) {
+		output.push({
+			type: "Analytics Engine Datasets",
+			entries: analytics_engine_datasets.map(({ binding, dataset }) => {
+				return {
+					key: binding,
+					value: dataset || binding,
```
Comment (penalosa) [inline packages/wrangler/src/config/index.ts:224]: `dataset ?? binding`

**Review by penalosa — APPROVED** [review]

File: packages/wrangler/src/type-generation.ts [hunk 4]
```diff
@@ -62,6 +62,12 @@ export async function generateTypes(
 		}
 	}
 
+	if (configToDTS.analytics_engine_datasets) {
+		for (const analyticsEngine of configToDTS.analytics_engine_datasets) {
+			envTypeStructure.push(`	${analyticsEngine.binding}: Dataset;`);
```
[the remark this answers is not in the document]
Comment (jimhawkridge) [reply packages/wrangler/src/type-generation.ts:67]: Good spot thanks. `AnalyticsEngineDataset` is correct. I'll update this one.

File: packages/wrangler/src/config/validation.ts (hunk 1 above)
Comment (jimhawkridge) [reply packages/wrangler/src/config/validation.ts:1992]: I copied this pattern from another binding and I think it is correct.
It allows you to have multiple errors in a binding and have both reported. 
I.e. this: `{ binding = 2, dataset = 4 }` correctly reports that both binding and dataset should be strings.

**Review by penalosa — APPROVED** [review]

**Pull request merged**